returnn/tf/layers/base.py

```diff
@@ -1755,8 +1759,12 @@ def custom_getter(getter, name, *args, **kwargs):
     assert name.startswith(abs_scope_prefix)
     param_name = name[len(abs_scope_prefix):]  # e.g. "W" (not "rec/W")
     if self.custom_func:
-      return self.custom_func(
+      variable = self.custom_func(
         base_layer=base_layer, reuse_layer=self.reuse_layer, name=param_name, getter=getter, full_name=name, **kwargs)
+      # The name of the variable created by custom_func might not match param_name.
+      # We store it here for LayerBase.add_param.
+      variable.RETURNN_layer_map_name = param_name
```

Comment:
I suggested `_RETURNN_layer_map_name` because this actually should be a map (dict). This cannot be the name directly, because this param might be used by multiple layers, under different names. So I thought about a mapping layer -> name. However, I'm now thinking, it might even be used by only one single layer under different names...

Reply:
Hm I see, I kind of thought that there are no unrelated calls to `custom_getter` before the corresponding `add_param` is called. That's of course not so nice (and probably breaks if e.g. a `custom_func` calls `get_variable` multiple times or so).
Maybe we can make it a layer -> list[name] dict (and then always use the last entry in `add_param`?) I don't really know a good solution.

Reply:
> and then always use the last entry ...

Such heuristics are never nice. They mostly work, but in some rare cases, or when we do sth more crazy, they suddenly fail. And then this is very annoying to debug. (Unfortunately, we have a couple of such heuristics, which have worked for the initial use case, but then later on failed, and caused quite some extra work. "Technical debt" is the keyword for this...)

But I think we don't need that here.

We can know when we are in the direct `get_variable` call from the layer. (Via `var_creation_scope`.)
And in there, we have the original name, and we can get the param, and we can directly do `self.params[name] = param`, and `add_param` doesn't need to do anything anymore.

And when `var_creation_scope` does not create a custom getter, this implies there will be no special handling. Then `add_param` can catch this.